### PR TITLE
Dictionaries - add support for migrations

### PR DESCRIPTION
**Pull request by edualonso**
Added support for migrations for dictionaries.

### CODE REVIEW

File: realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryMigrationTests.kt [hunk 1]
```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2020 Realm Inc.
```
Comment (clementetb) [inline realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryMigrationTests.kt:2]:
```suggestion
 * Copyright 2021 Realm Inc.
```

File: realm/realm-library/src/androidTest/kotlin/io/realm/GenericTester.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2020 Realm Inc.
```
Comment (clementetb) [inline realm/realm-library/src/androidTest/kotlin/io/realm/GenericTester.kt:2]:
```suggestion
 * Copyright 2021 Realm Inc.
```

File: realm/realm-library/src/androidTest/kotlin/io/realm/entities/DictionaryClass.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2020 Realm Inc.
```
Comment (clementetb) [inline realm/realm-library/src/androidTest/kotlin/io/realm/entities/DictionaryClass.kt:2]:
```suggestion
 * Copyright 2021 Realm Inc.
```

File: realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java [hunk 4]
```diff
@@ -525,7 +592,7 @@ Table getTable() {
     }
 
     static final Map<Class<?>, FieldMetaData> getSupportedSimpleFields() {
```
Comment (clementetb) [inline realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java:594]: The function name is misleading, it is just for lists. We could directly access the `SUPPORTED_LIST_SIMPLE_FIELDS ` as you do with `SUPPORTED_DICTIONARY_SIMPLE_FIELDS`.

**Review by clementetb — CHANGES_REQUESTED** [review]
Great job 👏👏👏 ! Added some minor cr.

**Comment by edualonso** [comment]
All comments have been addressed. I will merge this Monday morning If you guys have no more comments.

File: realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java [hunk 5]
```diff
@@ -48,32 +48,57 @@
...
-        m.put(byte.class, new FieldMetaData(RealmFieldType.INTEGER, RealmFieldType.INTEGER_LIST, false));
-        m.put(Byte.class, new FieldMetaData(RealmFieldType.INTEGER, RealmFieldType.INTEGER_LIST, true));
-        m.put(byte[].class, new FieldMetaData(RealmFieldType.BINARY, RealmFieldType.BINARY_LIST, true));
-        m.put(Date.class, new FieldMetaData(RealmFieldType.DATE, RealmFieldType.DATE_LIST, true));
-        m.put(ObjectId.class, new FieldMetaData(RealmFieldType.OBJECT_ID, RealmFieldType.OBJECT_ID_LIST, true));
-        m.put(Decimal128.class, new FieldMetaData(RealmFieldType.DECIMAL128, RealmFieldType.DECIMAL128_LIST, true));
-        m.put(UUID.class, new FieldMetaData(RealmFieldType.UUID, RealmFieldType.UUID_LIST, true));
-        m.put(Mixed.class, new FieldMetaData(RealmFieldType.MIXED, RealmFieldType.MIXED_LIST, true));
-        SUPPORTED_SIMPLE_FIELDS = Collections.unmodifiableMap(m);
+        Map<Class<?>, FieldMetaData> listMap = new HashMap<>();
```
Comment (rorbech) [inline realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java:55]: Again this looks like candidates for refactoring into the field type interface for #7320

**Review by rorbech — APPROVED** [review]
I am not familiar with the dynamic schema api, but LGTM.

**Pull request merged**